plugin/path_checkouts.go

```diff
@@ -5,11 +5,349 @@ import (
+				Callback: b.operationSetCheckIn,
+				Summary:  "Check service accounts in to the library.",
+			},
+		},
+		HelpSynopsis: `Check service accounts in to the library.`,
+	}
+}
+
+func (b *backend) operationSetCheckIn(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	return b.handleCheckIn(ctx, req, fieldData, false)
+}
+
+func (b *backend) pathSetManageCheckIn() *framework.Path {
+	return &framework.Path{
+		Pattern: libraryPrefix + "manage/" + framework.GenericNameRegex("name") + "/check-in$",
+		Fields: map[string]*framework.FieldSchema{
+			"name": {
+				Type:        framework.TypeLowerCaseString,
+				Description: "Name of the set.",
+				Required:    true,
+			},
+			"service_account_names": {
+				Type:        framework.TypeCommaStringSlice,
+				Description: "The username/logon name for the service accounts to check in.",
+			},
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.ReadOperation: &framework.PathOperation{
+				Callback: b.operationSetManageCheckIn,
+				Summary:  "Check service accounts in to the library.",
+			},
+		},
+		HelpSynopsis: `Force checking service accounts in to the library.`,
+	}
+}
+
+func (b *backend) operationSetManageCheckIn(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	return b.handleCheckIn(ctx, req, fieldData, true)
+}
+
+func (b *backend) handleCheckIn(ctx context.Context, req *logical.Request, fieldData *framework.FieldData, overrideCheckInEnforcement bool) (*logical.Response, error) {
+	setName := fieldData.Get("name").(string)
+
+	serviceAccountNamesRaw, serviceAccountNamesSent := fieldData.GetOk("service_account_names")
+	var serviceAccountNames []string
+	if serviceAccountNamesSent {
+		serviceAccountNames = serviceAccountNamesRaw.([]string)
+	}
+
+	set, err := readSet(ctx, req.Storage, setName)
+	if err != nil {
+		return nil, err
+	}
+	if set == nil {
+		return logical.ErrorResponse(`"%s" doesn't exist`, setName), nil
+	}
+
+	// If check-in enforcement is overridden or disabled at the set level, we should consider it disabled.
+	disableCheckInEnforcement := overrideCheckInEnforcement || set.DisableCheckInEnforcement
+
+	// Track the service accounts we check in so we can include it in our response.
+	var checkIns []string
+
+	// Build and validate a list of service account names that we will be checking in.
+	var cantCheckInErrs error
+	if len(serviceAccountNames) == 0 {
+		// It's okay if the caller doesn't tell us which service accounts they
+		// want to check in as long as they only have one checked out.
+		// We'll assume that's the one they want to check in.
+		toCheckIn := make(map[string]*locksutil.LockEntry)
+		for _, setServiceAccount := range set.ServiceAccountNames {
+			// This is called in a func to avoid calling defer in a loop.
+			if err := func() error {
+				lock := locksutil.LockForKey(b.checkOutLocks, setServiceAccount)
```

Comment:
Why use a lock per service account, instead of for the whole set? Aside from introducing a lot of complexity with closures and the defer/unlock handling, I'm pretty sure this can deadlock.  Service accounts names "foo" and "bar" may very well hash to the same lock, and in the `releaseLock = false` path this would deadlock.

Reply:
We _could_ do it that way. 

I was concerned about getting a deadlock too, so I wrote a race test where I contend for the same thing with 100 processes at once and I didn't get a deadlock. I think that's because the locks get released when the work is done, rather than continually held, letting things that are contending for the same one get through.

The downside of locking at a set level is that then, everything wanting to check out service accounts will have to do so synchronously. So, if there were 1,000 machines contending to check out 1,000 service accounts, they'd have to do it one-at-a-time rather than in parallel.

Reply:
Nevermind! Since we've decided to have check-ins and check-outs be transactional, I think locking at the set level is the way to go. That's mainly because no longer will I be acting on _this_ service account but not _that_ service account. It makes set-level locks more feasible and I think it'll simplify the code quite a lot.